### PR TITLE
Measure validation

### PR DESCRIPTION
**Pull request by sebbacon**
Adds a management command (`./manage.py import_measures --check`) which checks that measure SQL is valid.

Includes a step during fabric build which validates changed measures before building them, with the result that someone deploying a branch to staging will get an exception reported to `#technoise`.

Validation happens using BigQuery's `dryRun` option, which will raise exceptions on malformed SQL without incurring execution costs.

Note that this PR contains accidentally-merged commits from Rich which were then reverted, so commit-by-commit review will be hard; just check *Files changed*, which is not too big.

### CODE REVIEW

File: openprescribing/frontend/management/commands/import_measures.py [hunk 1]
```diff
@@ -31,6 +30,8 @@
 from common import utils
 from frontend.models import MeasureGlobal, MeasureValue, Measure, ImportLog
 
+import google
```
Comment (inglesp) [inline openprescribing/frontend/management/commands/import_measures.py]: Can you do `from google.api_core.exceptions import BadRequest` here?

File: openprescribing/frontend/management/commands/import_measures.py [hunk 2]
```diff
@@ -131,16 +163,27 @@ def parse_options(self, options):
...
 def parse_measures():
     """Deserialise JSON measures definition into dict
     """
     measures = OrderedDict()
-    fpath = os.path.dirname(__file__)
-    files = glob.glob(os.path.join(fpath, "./measure_definitions/*.json"))
-    for fname in sorted(files):
+    errors = []
+
+    for fname in get_measure_definition_files():
```
Comment (inglesp) [inline openprescribing/frontend/management/commands/import_measures.py]: These aren't filenames, they're paths, so I think `fname` isn't a great name.

File: openprescribing/frontend/tests/commands/test_import_measures.py [hunk 3]
```diff
@@ -31,15 +36,24 @@ def isclose(a, b, rel_tol=0.001, abs_tol=0.0):
...
     fpath = settings.REPO_ROOT
     fname = os.path.join(
-        fpath, ('openprescribing/frontend/tests/fixtures/'
-                'cerazette_measure.json'))
-    json_data = json.load(open(fname))
-    return {
-        'cerazette': json_data
-    }
+        fpath, ("openprescribing/frontend/tests/fixtures/measure_definitions/"
+                "{}.json".format(name)))
```
Comment (inglesp) [inline openprescribing/frontend/tests/commands/test_import_measures.py]: This is, to me, really oddly formatted.  I would much rather see us break the line limit here, than do this.

File: openprescribing/frontend/tests/commands/test_import_measures.py [hunk 4]
```diff
@@ -55,8 +69,8 @@ def parse_args(*opts_args):
     return cmd.parse_options(options.__dict__)
 
 
-@patch('frontend.management.commands.import_measures.parse_measures',
-       new=MagicMock(return_value=test_measures()))
+@patch('frontend.management.commands.import_measures.get_measure_definition_files',
```
Comment (inglesp) [inline openprescribing/frontend/tests/commands/test_import_measures.py]: May as well use `MODULE` here and in similar places below?

File: openprescribing/frontend/management/commands/import_measures.py [hunk 5]
```diff
@@ -131,16 +163,27 @@ def parse_options(self, options):
         return options
 
 
+def get_measure_definition_files():
```
Comment (inglesp) [inline openprescribing/frontend/management/commands/import_measures.py]: This doesn't return files, it returns paths.

File: openprescribing/gcutils/bigquery.py [hunk 6]
```diff
@@ -37,6 +40,25 @@ class BigQueryExportError(Exception):
     pass
 
 
+@contextmanager
+def exception_sql_printer(sql):
```
Comment (inglesp) [inline openprescribing/gcutils/bigquery.py]: Good idea!

**Comment by sebbacon** [comment]
You can make errors go to the originating channel (I think), but I wondered if that would be too noisy.... will give it a go.

**Pull request merged**